<a href="https://colab.research.google.com/github/idowujames/KPMG-Virtual-Data-Analytics-Internship/blob/main/KPMG_Virtual_Data_Analytics_Internship.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [47]:
import pandas as pd

# Data Quality Assessment

Access quality of the dataset provided by the client

In [48]:
# Func for reading in the datasets and skipping the first row

def read_data(data_path):
  return (pd.read_csv(data_path,skiprows=1))

In [49]:
# Reading in the datasets
transaction = read_data('https://raw.githubusercontent.com/idowujames/KPMG-Virtual-Data-Analytics-Internship/main/KPMG%20-%20Transactions.csv')
cust_address = read_data('https://raw.githubusercontent.com/idowujames/KPMG-Virtual-Data-Analytics-Internship/main/KPMG%20-%20CustomerAddress.csv')
cust_demographic = read_data('https://raw.githubusercontent.com/idowujames/KPMG-Virtual-Data-Analytics-Internship/main/KPMG%20-%20CustomerDemographic.csv')
cust_new = read_data('https://raw.githubusercontent.com/idowujames/KPMG-Virtual-Data-Analytics-Internship/main/KPMG%20-%20NewCustomerList.csv')

### Accessing The Transaction Data

In [50]:
transaction.head()

transaction_id  product_id  customer_id transaction_date online_order  \
0               1           2         2950        2/25/2017        False   
1               2           3         3120        5/21/2017         True   
2               3          37          402       10/16/2017        False   
3               4          88         3135        8/31/2017        False   
4               5          78          787        10/1/2017         True   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price standard_cost  product_first_sold_date  
0       71.49        $53.62                  41245.0  
1     2091.47       $388.92                  41701.0  
2     1793.43       $248.82                  36361.0  
3     1198.46       $381.10                  36145.0  
4     1765.30       $709.48                  42226.0

In [51]:
transaction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   transaction_id           20000 non-null  int64  
 1   product_id               20000 non-null  int64  
 2   customer_id              20000 non-null  int64  
 3   transaction_date         20000 non-null  object 
 4   online_order             19640 non-null  object 
 5   order_status             20000 non-null  object 
 6   brand                    19803 non-null  object 
 7   product_line             19803 non-null  object 
 8   product_class            19803 non-null  object 
 9   product_size             19803 non-null  object 
 10  list_price               20000 non-null  float64
 11  standard_cost            19803 non-null  object 
 12  product_first_sold_date  19803 non-null  float64
dtypes: float64(2), int64(3), object(8)
memory usage: 2.0+ MB


In [52]:
# Checking for missing values
transaction.isna().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [53]:
# Checking for duplicates entries accross the dataset
transaction[transaction.duplicated()]

Empty DataFrame
Columns: [transaction_id, product_id, customer_id, transaction_date, online_order, order_status, brand, product_line, product_class, product_size, list_price, standard_cost, product_first_sold_date]
Index: []

In [54]:
# Getting statical description of the number columns

transaction.select_dtypes('number').describe()

transaction_id   product_id   customer_id    list_price  \
count    20000.000000  20000.00000  20000.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050   1107.829449   
std       5773.647028     30.75359   1011.951046    582.825242   
min          1.000000      0.00000      1.000000     12.010000   
25%       5000.750000     18.00000    857.750000    575.270000   
50%      10000.500000     44.00000   1736.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000   1635.300000   
max      20000.000000    100.00000   5034.000000   2091.470000   

       product_first_sold_date  
count             19803.000000  
mean              38199.776549  
std                2875.201110  
min               33259.000000  
25%               35667.000000  
50%               38216.000000  
75%               40672.000000  
max               42710.000000

In [55]:
# Getting unique values in the categorical columns to confirm the correctness of values
columns = ['online_order','order_status','brand','product_line','product_class','product_size']
print("Unique Values - ")
for value in transaction[columns]:
  print(f"{value}: {transaction[value].unique()}")

Unique Values - 
online_order: [False True nan]
order_status: ['Approved' 'Cancelled']
brand: ['Solex' 'Trek Bicycles' 'OHM Cycles' 'Norco Bicycles' 'Giant Bicycles'
 'WeareA2B' nan]
product_line: ['Standard' 'Road' 'Mountain' 'Touring' nan]
product_class: ['medium' 'low' 'high' nan]
product_size: ['medium' 'large' 'small' nan]


In [56]:
# checking the transaction date to make sure there is no odd date
print(transaction['transaction_date'].min())
transaction['transaction_date'].max()

1/1/2017


'9/9/2017'

**Accessment**

- There are 13 columns

- There are missing values in the `online_order`, `brand`, `product_line`, `product_class`, `product_size`, `standard_cost` and `product_first_sold_date` columns

- The `product_first_sold_date` column is given as numbers instead of dates.

### Accessing The Customer Address Data

In [57]:
cust_address.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [58]:
cust_address.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [59]:
# Checking for duplicates entries accross the dataset
cust_address[cust_address.duplicated()]

Empty DataFrame
Columns: [customer_id, address, postcode, state, country, property_valuation]
Index: []

In [60]:
# Getting unique values in the categorical columns to confirm the correctness of values
columns = ['state','country']
print("Unique Values - ")
for value in cust_address[columns]:
  print(f"{value}: {cust_address[value].unique()}")

Unique Values - 
state: ['New South Wales' 'QLD' 'VIC' 'NSW' 'Victoria']
country: ['Australia']


**Accessment**

- There are 6 columns.

- No missing values.

- Postcode column should be string instead of number.


Seems everything else in the dataset is fine

### Accessing The Customer Demographics Data

In [61]:
cust_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases         DOB               job_title  \
0                                   93  1953-10-12     Executive Secretary   
1                                   81  1980-12-16  Administrative Officer   
2                                   61  1954-01-20      Recruiting Manager   
3                                   33  1961-10-03                     NaN   
4                                   56  1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                              1-Feb      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [62]:
cust_demographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          4000 non-null   int64  
 1   first_name                           4000 non-null   object 
 2   last_name                            3875 non-null   object 
 3   gender                               4000 non-null   object 
 4   past_3_years_bike_related_purchases  4000 non-null   int64  
 5   DOB                                  3913 non-null   object 
 6   job_title                            3494 non-null   object 
 7   job_industry_category                3344 non-null   object 
 8   wealth_segment                       4000 non-null   object 
 9   deceased_indicator                   4000 non-null   object 
 10  default                              3698 non-null   object 
 11  owns_car                      

In [63]:
# Checking for missing values
cust_demographic.isna().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

In [64]:
# Checking for duplicates entries accross the dataset
cust_demographic[cust_demographic.duplicated()]

Empty DataFrame
Columns: [customer_id, first_name, last_name, gender, past_3_years_bike_related_purchases, DOB, job_title, job_industry_category, wealth_segment, deceased_indicator, default, owns_car, tenure]
Index: []

In [66]:
# Getting unique values in the categorical columns to confirm the correctness of values
columns = ['gender','job_industry_category','wealth_segment','deceased_indicator','owns_car']
print("Unique Values - ")
for value in cust_demographic[columns]:
  print(f"{value}: {cust_demographic[value].unique()}")

Unique Values - 
gender: ['F' 'Male' 'Female' 'U' 'Femal' 'M']
job_industry_category: ['Health' 'Financial Services' 'Property' 'IT' nan 'Retail' 'Argiculture'
 'Manufacturing' 'Telecommunications' 'Entertainment']
wealth_segment: ['Mass Customer' 'Affluent Customer' 'High Net Worth']
deceased_indicator: ['N' 'Y']
owns_car: ['Yes' 'No']


In [83]:
# Converting the DOB column to date
cust_demographic['DOB'] = pd.to_datetime(cust_demographic['DOB'])

In [85]:
# checking the transaction date to make sure there is no odd date
print(cust_demographic['DOB'].min())
cust_demographic['DOB'].max()

1843-12-21 00:00:00


Timestamp('2002-03-11 00:00:00')

In [67]:
# Checking the number columns to make sure the values in it are within expected range
cust_demographic[['past_3_years_bike_related_purchases','tenure']].describe()

past_3_years_bike_related_purchases       tenure
count                          4000.000000  3913.000000
mean                             48.890000    10.657041
std                              28.715005     5.660146
min                               0.000000     1.000000
25%                              24.000000     6.000000
50%                              48.000000    11.000000
75%                              73.000000    15.000000
max                              99.000000    22.000000

**Accessment**

- There are 13 columns.

- There are missing values in the `last_name`, `DOB`, `job_title`, `job_industry_category`, `default` and `tenure`  columns

- Default column has some weird values in it.
- The `gender` column does not use uniform values to classify customer gender
- Some of the dates in the `DOB` have errors in it's values because the oldest person in the dataset was born in 1843

### Accessing The New Customer Data

In [68]:
cust_new.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

          DOB                   job_title job_industry_category  \
0  1957-07-12             General Manager         Manufacturing   
1  1970-03-22         Structural Engineer              Property   
2  1974-08-28      Senior Cost Accountant    Financial Services   
3  1979-01-28  Account Representative III         Manufacturing   
4  1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                 6.0        0.73      0.9125     1.140625     0.969531   
1                11.0        0.56      0.5600     0.700000     0.595000   
2                 5.0        0.44      0.4400     0.440000     0.440000   
3                 1.0        0.63      0.7875     0.787500     0.787500   
4                 9.0        0.45      0.4500     0.562500     0.562500   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [69]:
# Dropping the unamed columns as they are errors that occurred while converting dataset to csv
cust_new = cust_new.drop(columns=['Unnamed: 16','Unnamed: 17','Unnamed: 18','Unnamed: 19','Unnamed: 20'])
cust_new.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

          DOB                   job_title job_industry_category  \
0  1957-07-12             General Manager         Manufacturing   
1  1970-03-22         Structural Engineer              Property   
2  1974-08-28      Senior Cost Accountant    Financial Services   
3  1979-01-28  Account Representative III         Manufacturing   
4  1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  tenure              address  \
0      Mass Customer                  N      Yes      14     45 Shopko Center   
1      Mass Customer                  N       No      16    14 Mccormick Park   
2  Affluent Customer                  N       No      10  5 Colorado Crossing   
3  Affluent Customer                  N      Yes       5   207 Annamark Plaza   
4  Affluent Customer                  N       No      19    115 Montana Place   

   postcode state    country  property_valuation  Rank     Value  
0      4500   QLD  Australia                 6.0     1  1.718750  
1      2113   NSW  Australia                11.0     1  1.718750  
2      3505   VIC  Australia                 5.0     1  1.718750  
3      4814   QLD  Australia                 1.0     4  1.703125  
4      2093   NSW  Australia                 9.0     4  1.703125

In [70]:
cust_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 18 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   first_name                           1000 non-null   object 
 1   last_name                            971 non-null    object 
 2   gender                               1000 non-null   object 
 3   past_3_years_bike_related_purchases  1000 non-null   int64  
 4   DOB                                  983 non-null    object 
 5   job_title                            894 non-null    object 
 6   job_industry_category                835 non-null    object 
 7   wealth_segment                       1000 non-null   object 
 8   deceased_indicator                   1000 non-null   object 
 9   owns_car                             1000 non-null   object 
 10  tenure                               1000 non-null   int64  
 11  address                        

In [71]:
# Checking for duplicates entries accross the dataset
cust_new[cust_new.duplicated()]

Empty DataFrame
Columns: [first_name, last_name, gender, past_3_years_bike_related_purchases, DOB, job_title, job_industry_category, wealth_segment, deceased_indicator, owns_car, tenure, address, postcode, state, country, property_valuation, Rank, Value]
Index: []

In [73]:
# Getting unique values in the categorical columns to confirm the correctness of values
columns = ['gender','job_industry_category','wealth_segment','deceased_indicator','owns_car','state','country']
print("Unique Values - ")
for value in cust_new[columns]:
  print(f"{value}: {cust_new[value].unique()}")

Unique Values - 
gender: ['Male' 'Female' 'U']
job_industry_category: ['Manufacturing' 'Property' 'Financial Services' 'Entertainment' 'Retail'
 'IT' 'Telecommunications' 'Health' nan 'Argiculture']
wealth_segment: ['Mass Customer' 'Affluent Customer' 'High Net Worth']
deceased_indicator: ['N']
owns_car: ['Yes' 'No']
state: ['QLD' 'NSW' 'VIC']
country: ['Australia']


In [75]:
# Checking the number columns to make sure the values in it are within expected range
cust_new.describe()

past_3_years_bike_related_purchases       tenure     postcode  \
count                          1000.000000  1000.000000  1000.000000   
mean                             49.836000    11.388000  3019.227000   
std                              27.796686     5.037145   848.895767   
min                               0.000000     0.000000  2000.000000   
25%                              26.750000     7.000000  2209.000000   
50%                              51.000000    11.000000  2800.000000   
75%                              72.000000    15.000000  3845.500000   
max                              99.000000    22.000000  4879.000000   

       property_valuation         Rank        Value  
count         1000.000000  1000.000000  1000.000000  
mean             7.397000   498.819000     0.881714  
std              2.758804   288.810997     0.293525  
min              1.000000     1.000000     0.340000  
25%              6.000000   250.000000     0.649531  
50%              8.000000   500.000000     0.860000  
75%              9.000000   750.250000     1.075000  
max             12.000000  1000.000000     1.718750

In [87]:
# Converting the DOB column to date
cust_new['DOB'] = pd.to_datetime(cust_new['DOB'])

# checking the transaction date to make sure there is no odd date
print(cust_new['DOB'].min())
cust_new['DOB'].max()

1938-06-08 00:00:00


Timestamp('2002-02-27 00:00:00')

**Accessment**

- There are 18 columns.

- There are missing values in the `last_name`, `DOB`, `job_title` and`job_industry_category` columns.

- Default column has some weird values in it.
- Some of the dates in the `DOB` have errors in it's values because the oldest person in the dataset was born in 1938
